### DISCUSSION TITLE
How to create a submodule in the rust code?

### DISCUSSION THREAD

**Discussion by CMCDragonkai** [napi-rs/napi-rs#1379] [Q&A]
If I stick code into Rust modules, then use `mod x`, all of it gets exported at a top level. What's the right way to expose a submodule in JS? So I have to do `native.x.f`?

**Comment by Brooooooklyn** [comment]
you can add `#[napi]` to Rust `mod`:

```rust
#[napi]
pub fn foo() {}

#[napi(js_name = "bar")]
pub mod Bar {
  #[napi]
  pub fn baz() {}
}

```

**Reply by CMCDragonkai** [reply]
Does this work with just `mod filename`? That is my definitions are located in separate file. With `lib.rs` referencing them with `mod filename`.

**Reply by Brooooooklyn** [reply]
no, it doesn't work with `mod filename`

**Reply by CMCDragonkai** [reply]
So there's no way to export a module that sits in a separate file?

**Reply by Brooooooklyn** [reply]
Yes... supporting this feature would make code generation very complex (already complex for now). You can create JavaScript files for this purpose: https://github.com/Brooooooklyn/lzma/blob/main/lzma2.js
